Setup registry.dagger.io/engine proxy for ghcr.io/dagger/engine

gerhard opened the pull request:
This PR sets up https://registry.dagger.io based on the original registry-redirect fork.

It enables anyone to do this:

<img width="1005" alt="image" src="https://user-images.githubusercontent.com/3342/214274201-8a6eb98d-fecc-4729-9220-180ffab575d4.png">

### Why did the do this?

Regardless which container registry we end up using - `ghcr.io` today, but maybe `ecr.io` or `gcr.io` tomorrow - we want to have a small & simple front-end that we operate. This is especially important from the instrumentation perspective, since it's difficult (in some cases impossible) to understand how clients interact with the container registries listed above. Regardless which container registry we use, we want the same instrumentation interface, which this thin proxy is going to provide. This PR does not include this instrumentation, we intend to follow-up. This will most likely be JSON logs streamed to some back-end.

### What is happening behind the scenes?

```mermaid
flowchart LR
    client_can("👩🏽 🇨🇦")
    client_usa("🧔🏻 🇺🇸")
    client_arg("👶🏻 🇦🇷")
    client_uk("👨🏻‍🦰 🇬🇧")
    client_fr("👨🏻‍💻 🇫🇷")
    client_in("👨🏾 🇮🇳")
    
    subgraph Fly.io
        registry_redirect_usa("🇺🇸 registry.dagger.io/engine")
        registry_redirect_fr("🇫🇷 registry.dagger.io/engine")
        registry_redirect_jp("🇯🇵 registry.dagger.io/engine")
    end

    subgraph GitHub
        registry("🐙 ghcr.io/dagger/engine")
    end

    registry_redirect_usa --> registry
    registry_redirect_fr --> registry
    registry_redirect_jp --> registry

    client_usa --> registry_redirect_usa
    client_can --> registry_redirect_usa
    client_arg --> registry_redirect_usa
    client_uk --> registry_redirect_fr
    client_fr --> registry_redirect_fr
    client_in --> registry_redirect_jp
```

We are currently running 3 instances of `registry-redirect` for higher availability & lower edge latency:
1. 🇺🇸 San Jose
2. 🇫🇷 Paris
3. 🇯🇵 Tokyo

The above graph is a simplification. There are also **Edge** proxy instances running within the Fly.io network that serve clients directly. These are transparent to us, it's simply a Fly.io network optimisation. If you look at the world map in the screenshot below, you will notice that my `docker pull registry.dagger.io/engine:v0.3.9` above was actually serviced by the `LHR` edge proxy which connected to our closest registry-redirect instance running in `CDG` - 🇫🇷 Paris:

![image](https://user-images.githubusercontent.com/3342/214382839-2a56410d-74e2-493a-9eff-25ad9c595b99.png)

### Who is expected to interact with `registry.dagger.io` ?

For the most part, our SDKs: Go, Python & Node.js. Users are not expected to download our engine image directly, the SDKs will download it automatically, part of the provisioning step.

### How are we testing, publishing & deploying this?

It's all Dagger Go SDK wrapped in [Mage](https://magefile.org/). You will find the pipeline code in `magefiles/dagger.go`. 

This is what that looks like in practice:

![dagger-registry-2023-01-23-ci](https://user-images.githubusercontent.com/3342/214390075-0699bf9e-ed3f-44bb-9214-8298d248d137.gif)

This makes the GitHub Actions integration super simple. It's just a matter of running `go run main.go -w ../ all`

### What other commands did we run to set everything else up?

- `flyctl apps create dagger-registry-2023-01-23`
- `flyctl config save --app dagger-registry-2023-01-23`
- Make necessary edits to `fly.toml`
- Deploy app from local: `GITHUB_REF_NAME=main mage all`
- `flyctl regions add sjc cdg nrt`
- `flyctl scale 3`
- `flyctl regions delete lhr`
- Configure `registry.dagger.io` A & AAAA DNS record - `flyctl ips list`
- `flyctl certs create registry.dagger.io`

### Next steps

This is what I intend to continue with:

- Update `ghcr.io/dagger/engine` references in https://github.com/dagger/dagger with `registry.dagger.io/engine`
- Stream logs to Loki (Grafana Cloud)
- Visualise events
- Set up alerting

Commit:
Make it generic, customize & graceful

- Easy generic changes, e.g. comments, default values, etc.
- Customize variables & strings which did not make sense to customize at
  this stage
- Handle graceful shutdown via SIGINT

Follow-up improvement: extract specific values into a config file

Signed-off-by: Gerhard Lazu <gerhard@dagger.io>

Commit:
Daggerize & Fly

  cd magefiles
  go run main.go -w ../

  Targets:
    all        [lints, tests, auths], builds, publishes & deploys a new version of the app
    auth       Docker client with private registry
    build      binary artefact used in container image
    deploy     zero-downtime deploy container image
    lint       golangci-lint
    logs       stream app logs
    publish    container image to private registry
    test       go test

If you have mage - https://magefile.org/ - installed, then you can
simply run `mage` from the top-level.

There is also a top-level Fly config that captures the app deployment
config. Works on my machine™.

Signed-off-by: Gerhard Lazu <gerhard@dagger.io>

Commit:
Use Dagger in GitHub Actions

Only publish & deploy in CI, main branch.

Signed-off-by: Gerhard Lazu <gerhard@dagger.io>

sipsma commented on fly.toml:
```diff
@@ -0,0 +1,48 @@
+  protocol = "tcp"
+  script_checks = []
+  [services.concurrency]
+    hard_limit = 25
```
So is this a limit on the number of concurrent image pulls we will handle? What happens to clients when this limit gets hit (throttling error, call gets blocked, etc.)?

gerhard replied:
Yes. New requests will not finish the handshake and most likely result in timeouts.

Increasing this to 1000 sounds reasonable to me. I wouldn't go higher since I don't expect a single shared CPU to be able to service more than this and still offer a good QoS.

If we find ourselves hitting this limit, we should consider setting up autoscaling. FWIW https://community.fly.io/t/connection-count-limits/99

sipsma commented on magefiles/dagger.go:
```diff
@@ -0,0 +1,385 @@
+	}
+
+	if exitCode != 0 {
+		os.Exit(exitCode)
```
nit: os.Exit results in `defer`s made by callers to this func not running, so the dagger client won't be gracefully closed. I don't think this probably causes any problems right now though (the child session process should in theory handle its parent dying) so up to you if it's worth fixing right now

gerhard replied:
Yes, that is a good point, will handle it part of this.

sipsma reviewed (commented):
Minor comment+question, also it may be good to throw the list of manual `flyctl` commands you included in the PR description in a README here for easier access if needed in the future.

Commit:
Briefly explain the infra setup

Thanks for the nudge Erik!
https://github.com/dagger/registry-redirect/pull/1#pullrequestreview-1268313146

Signed-off-by: Gerhard Lazu <gerhard@dagger.io>

Commit:
Bump connection limits to 1000 & 800

Another good nudge!
https://github.com/dagger/registry-redirect/pull/1#discussion_r1085931658

Signed-off-by: Gerhard Lazu <gerhard@dagger.io>

Commit:
Replace os.Exit with panic & defer handleErr

So that caller defers are honored. Yup, also Erik!
https://github.com/dagger/registry-redirect/pull/1#discussion_r1085975449

Signed-off-by: Gerhard Lazu <gerhard@dagger.io>

gerhard commented:
Linted, tested, built, published & deployed in `2m 21s` 🚀

<img width="1439" alt="image" src="https://user-images.githubusercontent.com/3342/214670109-cf66ec4e-2de4-4b40-be7b-6b7b3d384b9f.png">

---

Now that this is merged, I will be looking into streaming logs to Grafana Cloud @grouville. Let's talk about https://vector.dev/ & https://github.com/superfly/fly-log-shipper tomorrow 🖖
